applications/serial_lte_modem/src/mqtt_c/slm_at_mqtt.c

```diff
@@ -217,41 +220,69 @@ static void mqtt_thread_fn(void *arg1, void *arg2, void *arg3)
 	while (true) {
 		if (!ctx.connected) {
 			LOG_WRN("MQTT disconnected");
+			err = 0;
 			break;
 		}
 		err = poll(&fds, 1, mqtt_keepalive_time_left(&client));
 		if (err < 0) {
 			LOG_ERR("ERROR: poll %d", errno);
+			err = -EIO;
 			break;
 		}
-		/* timeout or revent, send KEEPALIVE */
-		(void)mqtt_live(&client);
-
-		if ((fds.revents & POLLIN) == POLLIN) {
-			err = mqtt_input(&client);
-			if (err != 0) {
-				LOG_ERR("ERROR: mqtt_input %d", err);
-				mqtt_abort(&client);
-				break;
-			}
-		}
+		/* MQTT v3.1.1: Note that a Server is permitted to disconnect a Client that it
+		 * determines to be inactive or non-responsive at any time, regardless of the
+		 * Keep Alive value provided by that Client.
+		 */
 		if ((fds.revents & POLLERR) == POLLERR) {
 			LOG_ERR("POLLERR");
-			mqtt_abort(&client);
+			err = -ENETDOWN;
 			break;
 		}
 		if ((fds.revents & POLLHUP) == POLLHUP) {
 			LOG_ERR("POLLHUP");
-			mqtt_abort(&client);
+			err = -ECONNRESET;
 			break;
 		}
 		if ((fds.revents & POLLNVAL) == POLLNVAL) {
 			LOG_ERR("POLLNVAL");
-			mqtt_abort(&client);
+			err = -ENOTCONN;
 			break;
 		}
+		if ((fds.revents & POLLIN) == POLLIN) {
+			err = mqtt_input(&client);
+			if (err != 0) {
+				LOG_ERR("ERROR: mqtt_input %d", err);
+				break;
+			}
+			/* MQTT v3.1.1: If a Client does not receive a PINGRESP Packet within a
+			 *  reasonable amount of time after it has sent a PINGREQ, it SHOULD close
+			 *  the Network Connection to the Server.
+			 */
+			if (client.unacked_ping > 0) {
```

Comment:
How does this work for scenario where we send the ping request, but at the same time the server sends us anoher message? Wont we then hit this and connect erronously?

Reply:
The spec says client could disconnect if no response for mqtt ping

Reply:
This works well for a scenario where we ping and after some time receive a PUB-message, but if we receive a PUB-message when we have just send the ping, we will hit the POLLIN branch and now have unacked ping. Then we will disconnect erroneously.

This should have some kind of time treshold or other mechanism to prevent above scenario.

Reply:
Discussed this with @junqingzou, we currently have a similar implementation in nrf_cloud. And while the implementation does not work well in this particular case, it is not critical as the connection can terminate for multitude of reasons and there should either be user re-trial or re-trial logic in case of usage as SW component.

If this becomes problematic, the correct way to fix this, would be to note that  we have sent ping req, lower the poll timeout and check whether we have received ping resp during the timeout. It is not straightforward solution as there can again be other packets coming through this time. 

It is possible that this cannot be improved without a separate timer or having time as a concept in here.

But I think that we can go with this for now. A quick fix in possible problem case is to set the unacked_ping > 1.